File: BOJ9012.cpp
#include<iostream>
#include<string>
#include<stack>

using namespace std;

int main() {
	int t;
	cin >> t;
	stack<char> s;
	string str;
	int strlen;
	while (t--) {
		cin >> str;
		strlen = str.length();
		for (int i = 0; i < strlen; i++) {
			if (str[i] == '(') {
				s.push('(');
			}
			else if (str[i] == ')') {
				if (s.empty() == true) {
					s.push('x');
					break;
				}
				else
					s.pop();
			}
		}
		if (s.empty() == true)
			cout << "YES\n";
		else {
			cout << "NO\n";
		}
		while (!s.empty()) s.pop();
	}
}

File: BOJ9012.java
package BOJ;

import java.util.Scanner;
import java.util.Stack;

public class Main9012 {
	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);

		int n = scan.nextInt();
		scan.nextLine();
		
		Stack<Integer> stack = new Stack<Integer>();
		
		while (n-- > 0) {
			String ps = scan.nextLine();
			String[] str = ps.split("");
			int psLen = ps.length();
			for (int i = 0; i < psLen; i++) {
				if (str[i].equals("(")) {
					stack.push(1);
				} else {
					if (stack.empty()) {
						stack.push(0);
						break;
					} else
						stack.pop();
				}
			}

			if(stack.empty())
				System.out.println("YES");
			else
				System.out.println("NO");
			
			while (!stack.empty()) {
				stack.pop();
			}
		}
		scan.close();
	}
}

//// 백준 정답
// import java.util.*;
// import java.math.*;

// public class Main {
//     public static String isValid(String s) {
//         s = s.trim();
//         int n = s.length();
//         int cnt = 0;
//         for (int i=0; i<n; i++) {
//             if (s.charAt(i) == '(') {
//                 cnt += 1;
//             } else {
//                 cnt -= 1;
//             }
//             if (cnt < 0) {
//                 return "NO";
//             }
//         }
//         if (cnt == 0) {
//             return "YES";
//         } else {
//             return "NO";
//         }
//     }
//     public static void main(String args[]) {
//         Scanner sc = new Scanner(System.in);
//         int n = sc.nextInt();
//         sc.nextLine();
//         while (n-- > 0) {
//             System.out.println(isValid(sc.nextLine()));
//         }
//     }
// }
